### PR TITLE
[Bug](topn opt) Fix Two-Phase read when some rowset swept

### PR DESCRIPTION
**Pull request by eldenmoon**
If this is a Two-Phase read query, and we need to delay the release of Rowset by row->update_delayed_expired_timestamp() to expand the lifespan of rowsets. This is necessary to avoid data loss during the second phase reading, where some stale rowsets may be swept and result in missing data.
    
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

File: be/src/runtime/query_context.h [hunk 1]
```diff
@@ -222,6 +228,14 @@ class QueryContext {
     taskgroup::TaskGroupPtr _task_group;
     std::unique_ptr<RuntimeFilterMgr> _runtime_filter_mgr;
     const TQueryOptions _query_options;
+
+    // When querying, certain structures need to be delayed in their release.
+    // Placing these structures in _reference_holder can extend their lifespan until the query is completed,
+    // and they are automatically released. For example, in two-phase delayed materialization queries,
+    // Rowsets need to be released after the query is finished to avoid premature release during the query process,
+    // which could lead to query exceptions.
+    std::vector<std::shared_ptr<void>> _reference_holder;
```
Comment (yiguolei) [inline be/src/runtime/query_context.h:237]: Do not use a very abstract structure. I think you could use std::shared_ptr<Rowset> and then the maintainer could understand very clear that the query context hold some rowset. Or there maybe some memory leak for example, if some other people add olap scan node to query context. The olap scan node maybe a shared ptr, it owns query context and query context owns olap scannode .

File: be/src/service/internal_service.cpp [hunk 2]
```diff
@@ -1512,11 +1512,18 @@ Status PInternalServiceImpl::_multi_get(const PMultiGetRequest& request,
...
             continue;
         }
+        // Get Rowset from either tablet or unused rowsets, since this rowset maybe expired and swept.
+        // But we ensured it's rowset is not released when init Tablet reader param, by QueryContext::hold_reference
         BetaRowsetSharedPtr rowset =
                 std::static_pointer_cast<BetaRowset>(tablet->get_rowset(rowset_id));
         if (!rowset) {
-            LOG(INFO) << "no such rowset " << rowset_id;
-            continue;
+            std::optional<RowsetSharedPtr> from_unused_rowset =
```
Comment (yiguolei) [inline be/src/service/internal_service.cpp:1520]: If it is hold in query context, why not get it from query context?

File: be/src/olap/storage_engine.cpp [hunk 3]
```diff
@@ -1155,10 +1155,15 @@ Status StorageEngine::execute_task(EngineTask* task) {
 }
 
 // check whether any unused rowsets's id equal to rowset_id
-bool StorageEngine::check_rowset_id_in_unused_rowsets(const RowsetId& rowset_id) {
+std::optional<RowsetSharedPtr> StorageEngine::get_rowset_in_unused_rowsets(
```
Comment (yiguolei) [inline be/src/olap/storage_engine.cpp:1158]: I think its very bad to get a rowset from gc space. GC is a standalone service. Other logic should not depend on it. The rowset is already hold in query context, I think you should get it from query context.

File: be/src/runtime/query_context.h (hunk 1 above)
Comment (eldenmoon) [reply be/src/runtime/query_context.h]: done

File: be/src/service/internal_service.cpp (hunk 2 above)
Comment (eldenmoon) [reply be/src/service/internal_service.cpp]: done

File: be/src/olap/storage_engine.cpp (hunk 3 above)
Comment (eldenmoon) [reply be/src/olap/storage_engine.cpp]: done

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**